File: pyleetcode/easy/single_number.py
"""136. Single Number

https://leetcode.com/problems/single-number/
"""

from typing import Optional, List, Any  # noqa: F401


class Solution:

    def singleNumber(self, nums: List[int]) -> int:
        result = 0
        for n in nums:
            result ^= n
        return result

    def process(self, *args, **kwargs):  # noqa: N802
        return self.singleNumber(*args, **kwargs)

File: rsleetcode/src/easy/single_number.rs
//! 136. Single Number
//!
//! https://leetcode.com/problems/single-number/

pub struct Solution {}

impl Solution {
    pub fn single_number(nums: Vec<i32>) -> i32 {
        nums.iter().fold(0, |acc, n| acc ^ n)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_1() {
        assert_eq!(Solution::single_number(vec![2, 2, 1]), 1)
    }

    #[test]
    fn test_2() {
        assert_eq!(Solution::single_number(vec![4, 1, 2, 1, 2]), 4)
    }

    #[test]
    fn test_3() {
        assert_eq!(Solution::single_number(vec![1]), 1)
    }
}
